### PR TITLE
Implement rotary embeddings

### PR DESCRIPTION
**Pull request by thomasw21**
Fixes: https://github.com/bigscience-workshop/Megatron-DeepSpeed/issues/6

## Description

Mostly copy work already done from EleutherAI (I've left a link at the top of `positional_embeddings.py`).

It slightly differs from the paper implementation. In particular, it permutes the ordering of the dimension in order to have an easier to compute rotary matrix multiplication.

Also:
 - I've slightly modified the caching mechanism for positional embeddings compared to EuleutherAI as it seems to me there's a very easy way to have a better caching mechanism.

TODO:
 - [x] Implement rotary
 - [x] Be backward compatible with absolute position embeddings (we don't handle it for vit ...)
 - [x] Test that the code works.
 - [ ] Check loss goes down.
 
 ## Changes

- To run the option, you have to use `--position-embedding-type rotary`
- Previous absolute embeddings is used using the following arguments `--position-embedding-type absolute --max-absolute-position-embeddings 512`
- We move `enum.py` from `megatron/model/` to `megatron/`. The pratical reason is because it provoked circular dependency due to `megatron/model/__init__.py`. The "good" reason, is IMO this shouldn't live in model as the whole project should be able to access those enums. I can revert back and put my custom enum in `arguments.py` if you disagree with me.

### CODE REVIEW

File: megatron/arguments.py [hunk 1]
```diff
@@ -199,8 +202,7 @@ def parse_args(extra_args_provider=None, defaults={},
                 'and lr-warmup-samples'
 
     # Check required arguments.
-    required_args = ['num_layers', 'hidden_size', 'num_attention_heads',
-                     'max_position_embeddings']
+    required_args = ['num_layers', 'hidden_size', 'num_attention_heads', 'position_embedding_type']
```
Comment (TevenLeScao) [inline megatron/arguments.py:205]: can we make embedding type absolute by default ? keeps backwards compatibility, also this way we don't have to retroactively change every single script!

File: megatron/arguments.py [hunk 2]
```diff
@@ -219,10 +221,15 @@ def parse_args(extra_args_provider=None, defaults={},
         assert args.encoder_seq_length is not None
         args.seq_length = args.encoder_seq_length
 
-    if args.seq_length is not None:
-        assert args.max_position_embeddings >= args.seq_length
-    if args.decoder_seq_length is not None:
-        assert args.max_position_embeddings >= args.decoder_seq_length
+    if args.position_embedding_type == PositionEmbeddingType.absolute:
```
Comment (TevenLeScao) [inline megatron/arguments.py:224]: think that thanks to this block we can just remove both `max_position_embeddings` and `embedding_type` from the required args

File: megatron/arguments.py [hunk 3]
```diff
@@ -279,7 +286,7 @@ def _add_network_size_args(parser):
                        'attention. This is set to '
                        '   args.hidden_size // args.num_attention_heads '
                        'if not provided.')
-    group.add_argument('--max-position-embeddings', type=int, default=None,
+    group.add_argument('--max-absolute-position-embeddings', type=int, default=None,
```
Comment (TevenLeScao) [inline megatron/arguments.py:289]: not a fan of changing the argument name; let's keep it the same to have backward compatibility, it's fine if it only applies to one embedding type

File: megatron/model/fused_softmax.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
 # limitations under the License.
 
 import torch
-from megatron.model.enums import AttnMaskType
+from megatron.enums import AttnMaskType
```
Comment (TevenLeScao) [inline megatron/model/fused_softmax.py:17]: This move surprises me. Did you move `enums` up a folder and if so, why?

**Review by TevenLeScao — CHANGES_REQUESTED** [review]

File: megatron/model/fused_softmax.py (hunk 4 above)
Comment (thomasw21) [reply megatron/model/fused_softmax.py:17]: Basically this caused circular dependency, when I added my enum in that file, and imported it in `arguments.py`. The reason why, is when you import `from megatron.model.enums import AttnMaskType` you'd execute `megatron/model/__init__.py` which means you're importing a bunch of code, some of which import `arguments.py`.
https://stackoverflow.com/questions/24302754/python-submodule-imports-using-init-py/24303380

In order to remove this dependency I've moved enums outside model, as it's safe to say that importing enums should not be linked to model. Please see the section `Changes` in the PR description.

File: megatron/arguments.py (hunk 1 above)
Comment (thomasw21) [reply megatron/arguments.py]: `position_embedding_type` is kept as it can't be None.

File: megatron/arguments.py (hunk 2 above)
Comment (thomasw21) [reply megatron/arguments.py:224]: See previous remark

**Review by TevenLeScao — APPROVED** [review]

File: megatron/model/positional_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+# rotary pos emb helpers:
+
+def rotate_half(x):
+    x1, x2 = x[..., :x.shape[-1] // 2], x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=x1.ndim - 1)  # dim=-1 triggers a bug in earlier torch versions
+
+
+@torch.jit.script
+def apply_rotary_pos_emb(q, k, cos, sin, offset: int = 0):
```
[the remark this answers is not in the document]
Comment (thomasw21) [reply megatron/model/positional_embeddings.py:44]: So I'm not familiar with `torch.jit.script` but I guess it compiles it? I admit that the code was copy pasted from EleutherAI with some little modifications. I'm okay with removing this helper if you want.

**Comment by stas00** [comment]
This broke the possibility to load our checkpoints into Megatron-LM or original Megatron-Deepspeed because they don't have `megatron.enum` and `torch.load` breaks.

Would it be possible to restore it to how it was?
